### DISCUSSION TITLE
Better interface code-deduplication features

### DISCUSSION THREAD

**Discussion by urkle** [async-graphql/async-graphql#1134] [Ideas]
I am using interfaces and currently there is a lot of duplication needed to implement them in each object it would be nice if we could find some clean ways to de-duplicate code.

There have been suggestions of using generics #978

Other ideas would be using traits (E.g. have "shared" implementations in a trait). Juniper implements interfaces via traits, however that may get tricky with async (hopefully async-trait crate will provide enough to deal with that)

And another would be to support "flatten" on the interface definition.

In my (current) use-case I have a nested object within my two objects that implement an interface so it would be super nice if I could just implement a single method and say 'flatten it'.  However when I do that, the interface code balks as the individual methods are not implemented and there is no way to tell the interface code "hey I'm flattening this!"

@sunli829 what are your thoughts on any of these approaches? (based on your previous experience).

**Comment by sunli829** [comment]
I have tried using traits to implement the interface, but failed, so I am basically satisfied with the current design. 😂
